app/mbm/static/js/app.js

```diff
@@ -0,0 +1,370 @@
+    this.routeListUrl = routeListUrl
+    this.routeUrl = routeUrl
+
+    this.routeLayer = null
+    this.allRoutesLayer = null
+    this.markers = { 'source': null, 'target': null }
+
+    // Start the app once the DOM is ready
+    document.addEventListener('DOMContentLoaded', this.start.bind(this))
+  }
+
+  start() {
+    // Create the leaflet map
+    this.map = this.createMap()
+
+    // Store references to DOM elements we'll need
+    const $directionsForm = $('#input-elements')
+    const sourceCoordsElem = document.getElementById('source')
+    const sourceTextInput = document.getElementById('source_text')
+
+    const targetCoordsElem = document.getElementById('target')
+    const targetTextInput = document.getElementById('target_text')
+
+    // This uses the same keys as the `markers` object for convenience
+    // in the code below
+    this.directionsFormElements = {
+      source: {
+        coords: sourceCoordsElem,
+        input: sourceTextInput,
+        autocomplete: null
+      },
+      target: {
+        coords: targetCoordsElem,
+        input: targetTextInput,
+        autocomplete: null
+      }
+    }
+
+    this.$routeEstimate = $('#route-estimate')
+
+    // Setup interactive tooltip elements (via jQuery UI)
+    $('[data-toggle="tooltip"]').tooltip()
+
+    this.gpsLocationString = 'My position'
+
+    const isMobileScreen = $(window).outerWidth() <= 768
+
+    // Make sure the map always fits the full height of the screen
+    $(window).resize(() => {
+      var windowHeight = $(window).innerHeight()
+      var offsetTop = $('.navbar')[0].offsetHeight
+      // Add controls to the top offset on mobile screens, where they merge
+      // with the navbar
+      if (isMobileScreen) { offsetTop += $('#controls-container')[0].offsetHeight }
+      var mapHeight = windowHeight - offsetTop
+      $('#map').css('height', mapHeight)
+    }).resize()
+
+    // Recalculate map size when controls are toggled
+    $directionsForm.on('shown.bs.collapse hidden.bs.collapse', function (e) {
+      $(window).resize()
+    })
+
+    this.createLegend().addTo(this.map)
+
+    // Allow users to name and save their own locations by double clicking on the map
+    this.userLocationsCheckbox = document.getElementById('enable-user-locations')
+    this.handleLocationsCheckboxChange = (event) => {
+      this.preferences.userLocationsEnabled = event.target.checked
+      saveUserPreferences(this.preferences)
+
+      if (event.target.checked && !this.userLocations) {
+        this.userLocations = new UserLocations(this)
+      } else {
+        this.userLocations.unmount()
+        this.userLocations = null
+      }
+    }
+    this.userLocationsCheckbox.addEventListener('change', this.handleLocationsCheckboxChange)
+
+    this.preferences = getUserPreferences()
+    if (this.preferences.userLocationsEnabled) {
+      this.userLocationsCheckbox.checked = true
+      this.userLocationsCheckbox.dispatchEvent(new Event('change'))
+    }
```

Comment:
>it might be worthwhile for us to eventually factor all the different setup steps into dedicated helper methods to aid in readability

I definitely think we should do this for readability. I started moving towards that but was trying to keep the scope of this from getting too large. Basically I just wanted to do the minimum organization so that these new features don't all have to go in one file, but this App class is way too big and definitely needs to be broken up more.